### PR TITLE
Add Ruby 3.1 and Rails 6.1 and 7 to the CI matrix

### PR DESCRIPTION
**Pull request by petergoldstein**
This PR adds explicit gemfiles for Rails 6.1 and Rails 7.  It updates the Ruby 3.1 configuration to run against these two new entries as well as against Rails edge.

### CODE REVIEW

File: gemfiles/rails_6_1.gemfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+eval_gemfile "../Gemfile"
+
+if Gem::Version.new(RUBY_VERSION.dup) >= Gem::Version.new("3.1")
```
Comment (Mangara) [inline gemfiles/rails_6_1.gemfile:5]: Can you explain why this is necessary?

File: gemfiles/rails_6_1.gemfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  gem "net-imap", require: false
+  gem "net-pop", require: false
+  gem "net-smtp", require: false
```
Comment (Mangara) [inline gemfiles/rails_6_1.gemfile:8]: Should we specify versions for these gems?

**Review by Mangara — COMMENTED** [review]
Thanks for the PR! We should definitely do this.
I just have some questions about the conditional gems for the 6.1 gemfile.

File: gemfiles/rails_6_1.gemfile (hunk 1 above)
Comment (petergoldstein) [reply gemfiles/rails_6_1.gemfile:5]: The `mail` gem has an issue discussing the need here - https://github.com/mikel/mail/pull/1439 .  And the Rails team discusses it here - https://github.com/rails/rails/pull/44600 .

These gems are not included by default in Ruby 3.1, so they need to be explicitly included.  Attempting to explicitly include them in older Rubies causes issues, so the invoking Gemfile is supposed to explicitly include them.

File: gemfiles/rails_6_1.gemfile (hunk 2 above)
Comment (petergoldstein) [reply gemfiles/rails_6_1.gemfile:8]: I don't think it's required.  I haven't seen any of the fixes for this issue in various repos add versions.

**Comment by Mangara** [comment]
I want to see a CI run for this PR before merging. I think we had an issue in our workflow setup that prevented it from triggering a CI run. That should be corrected after merging https://github.com/Shopify/job-iteration/pull/195, so I'm going to lose and re-open this PR to try and trigger a CI run.

**Review by Mangara — APPROVED** [review]

**Pull request closed without merging**